### DISCUSSION TITLE
Cookie not sent to api

### DISCUSSION THREAD

**Discussion by farahats9** [nuxt/nuxt#21438] [Questions]
I am trying to set the header to add the cookie but it doesn't work when doing a `post` request to the api

```typescript
const config = useRuntimeConfig();

const onSubmit = async () => {
    await useFetch(`/some/api/endpoint`, {
        baseURL: config.public.API_BASE_URL,
        method: 'POST',
        body: form.value,
        headers: { 'X-CSRF-Token': csrfData.csrf_token, credentials: 'include', Cookie: document.cookie }
    })
}
```

I have also tried `useRequestHeaders(['cookie'])` as in [this](https://nuxt.com/docs/getting-started/data-fetching#pass-client-headers-to-the-api) example but it didn't work.

**Comment by maou-shonen** [comment]
[useCookie](https://nuxt.com/docs/api/composables/use-cookie#usecookie)

```ts
const fooCookie = useCookie('foo')
fooCookie.value = 'bar'

// your fetch
useFetch('/api/foo', {
  server: false
})
```

**Reply by farahats9** [reply]
Thank you for trying to help. This didn't work for me at all. Even though the cookie is there in the application tab in dev tools

**Reply by maou-shonen** [reply]
This example does indeed work.
If it doesn't suit your needs,
I believe you should provide more information or a minimal reproducible in order to identify the issue.

**Reply by maou-shonen** [reply]
I noticed your other question.
perhaps your cookie is not passing through CORS?

**Reply by farahats9** [reply]
I am not sure what else I need to provide so here is a full page I made and tested:

```vue
<template>
    <div>
        <h1>test page for cookie</h1>
        <button @click="onSubmit">Submit</button>
    </div>
</template>

<script setup lang="ts">

const form = ref({
    name: 'abc',
    email: 'test@test.com',
    message: '123',
    token: 'some token'
})
const event = useRequestEvent()
const config = useRuntimeConfig();

const csrfData = await fetchWithCookie(event, `/api/csrf`, {
    server: true, key: 'csrf',
})
// there is no issue above here, request is made to api endpoint and returns set-cookie header and the cookie is set

const csrfCookie = useCookie('fastapi-csrf-token')

const onSubmit = async () => {
    console.log(csrfCookie.value)  // the token prints here to console correctly so it is available
    const headers = useRequestHeaders(['cookie'])  // tried with and without this but nothing worked
    await useFetch(`/api/contact-us`, {
        baseURL: config.public.API_BASE_URL,
        method: 'POST',
        body: form.value,
        headers: { 'X-CSRF-Token': csrfData.csrf_token, ...headers },
        server: false,
    })
}

</script>
```

here is how the request headers look in the browser:
![image](https://github.com/nuxt/nuxt/assets/463359/504b6691-db17-43c5-a330-6c640d6a66b0)
you will notice the `X-CSRF-Token` header is correctly set but not the cookies, I tried setting the `Cookie` header manually even and nothing worked.

and finally for your comment about CORS I am pretty sure I set it correctly on my backned
```python
        allow_origins="*",
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
```

I also tried changing the cookie `SameSite` policy from strict to lax to none and still no success.

EDIT: I am using nuxt 3.5.0 and nitro 2.4.1 if that makes any difference.

**Reply by stafyniaksacha** [reply]
You cannot use `Access-Control-Allow-Origin=*` with `Access-Control-Allow-Credentials=true`, you must specify allowed domains: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin#directives

> Attempting to use the wildcard with credentials [results in an error](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials).

**Reply by logotip4ik** [reply]
In current situation you should use `$fetch`, instead of `useFetch`. Good rule of thumb to use `$fetch` with everything that requires user interaction and `useFetch` in top level context without user interaction. Also you can check that `useRequestHeaders` returns `undefined` since this composable should be only called withing the server, you can not have access to request event on the client side. In your initial example you have used `document.cookies` and if this does not work, i would say that cookie you need is `httpOnly` so it cannot be accessed within client side js, and if cookie domain is not set correctly it would not end up in request to your backend.
